crypto/ec/ec_evp_lib.c

```diff
@@ -0,0 +1,444 @@
+
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_EXCHANGE_PARAM_KDF_TYPE,
+                                            /*
+                                             * Cast away the const. This is read
+                                             * only so should be safe
+                                             */
+                                            (char *)kdf_type, strlen(kdf_type));
+    *p++ = OSSL_PARAM_construct_end();
+
+    if ((ret = evp_pkey_ctx_set_params_strict(ctx, params)) == -2 ) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
+
+    return ret;
+}
+
+int EVP_PKEY_CTX_get_ecdh_kdf_type(EVP_PKEY_CTX *ctx)
+{
+    int ret;
+    /* 80 should be big enough */
+    char kdf_type[80];
+    OSSL_PARAM params[2], *p = params;
+
+    if ( 1 != (ret = evp_pkey_ctx_getset_ecdh_param_checks(ctx)))
+        return ret;
+
+    /* TODO(3.0): Remove this eventually when no more legacy */
+    if (ctx->op.kex.exchprovctx == NULL)
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_TYPE, -2, NULL);
+
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_EXCHANGE_PARAM_KDF_TYPE,
+                                            kdf_type, sizeof(kdf_type));
+    *p++ = OSSL_PARAM_construct_end();
+
+    if ((ret = evp_pkey_ctx_get_params_strict(ctx, params)) == -2 ) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    } else if (ret != 1) {
+        return -1;
+    }
+
+    if (kdf_type[0] == '\0')
+        return EVP_PKEY_ECDH_KDF_NONE;
+    else if (strcmp(kdf_type, OSSL_KDF_NAME_X963KDF) == 0)
+        return EVP_PKEY_ECDH_KDF_X9_63;
+
+    return -1;
+}
+
+int EVP_PKEY_CTX_set_ecdh_kdf_md(EVP_PKEY_CTX *ctx, const EVP_MD *md)
+{
+    int ret;
+    OSSL_PARAM params[2], *p = params;
+    const char *md_name = NULL;
+
+    if ( 1 != (ret = evp_pkey_ctx_getset_ecdh_param_checks(ctx)))
+        return ret;
+
+    /* TODO(3.0): Remove this eventually when no more legacy */
+    if (ctx->op.kex.exchprovctx == NULL)
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_MD, 0, (void *)(md));
+
+    md_name = (md == NULL) ? "" : EVP_MD_name(md);
+
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_EXCHANGE_PARAM_KDF_DIGEST,
+                                            /*
+                                             * Cast away the const. This is read
+                                             * only so should be safe
+                                             */
+                                            (char *)md_name,
+                                            strlen(md_name)+1);
```

Comment:
To include the NUL byte, as `OSSL_PARAM`'s data_size i the size of the data regardless of type, and the NUL byte is very much part of the data of a string, at least in C.

However, if you simply pass 0, `OSSL_PARAM_construct_utf8_string` will calculate the size for you:

https://github.com/openssl/openssl/blob/7f293d9f3b5cee4f4b15624fff15a45e0517334f/crypto/params.c#L849-L850

Reply:
I'd suggest just leaving this as 0 and letting the construct function do the strlen.

Reply:
I mentioned it because there is another utf8_string instance which doesn't do this (see the one with OSSL_EXCHANGE_PARAM_KDF_TYPE)

Reply:
I did it this way for conformity with:

https://github.com/openssl/openssl/blob/450d12c825cc9016e5e8990423fa7ffdb843a1f0/crypto/rsa/rsa_lib.c#L845-L865

and 

https://github.com/openssl/openssl/blob/450d12c825cc9016e5e8990423fa7ffdb843a1f0/crypto/rsa/rsa_lib.c#L977-L993

Reply:
@mattcaswell, I would argue that the `OSSL_EXCHANGE_PARAM_KDF_TYPE` is wrong. However, you get away with it because the receiving end doesn't always check the length (it's arguably redundant)

Reply:
Oh wait, I might take that back

Reply:
So no, the handling of `OSSL_EXCHANGE_PARAM_KDF_TYPE` above *is* correct.

It seems that you guys need to be reminded that:
1.  There is a fundamental difference between `strlen` and `sizeof`
2.  There is a fundamental difference between *setting* and *getting* parameters.

Reply:
So here, should I keep it like this for uniformity?

We can have a separate PR where we go ahead and follow @paulidale advice (use `0` as the last argument) and apply this in all the places where we are currently following the `strlen(buf)+1` pattern.

Reply:
I frankly don't care much, as long as you get it right :wink:

Do note, though, that the 0 argument only works for utf8_string.

Reply:
Ok, so my preference goes to keeping it as it is right now, and then do a PR that applies @paulidale suggestions in all the proper places rather than adding that pass to this PR.

Reply:
> So no, the handling of OSSL_EXCHANGE_PARAM_KDF_TYPE above is correct.
>
> It seems that you guys need to be reminded that:
>
> There is a fundamental difference between strlen and sizeof
> There is a fundamental difference between setting and getting parameters.

@levitte I was referring to the this line:

````
    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_EXCHANGE_PARAM_KDF_TYPE,
                                            /*
                                             * Cast away the const. This is read
                                             * only so should be safe
                                             */
                                            (char *)kdf_type, strlen(kdf_type));
````

Which is in `EVP_PKEY_CTX_set_ecdh_kdf_type` above: its a setter (just like this instance is) and uses strlen (just like this instance). But this instance has "+1" but `EVP_PKEY_CTX_set_ecdh_kdf_type` does not

Reply:
I'll have a look around...